### PR TITLE
Replace some sizeof-s with target specific constants

### PR DESCRIPTION
**Pull request by echesakovMSFT**
The goal of this PR is to identify all places in VM using sizeof operator that should be made target-specific in order to support cross-bitness scenario #16513. <s>The proposal is to introduce CrossBitness_SizeOf operator to make such places *explicit*.

CrossBitness_SizeOf operator satisfies the following conditions:
1. CrossBitness_SizeOf(X) == sizeof(X) in non-crossbitness scenarios;
2. CrossBitness_SizeOf(X) == sizeof(Y) where Y is a "fake" structure created to describe layout of X on the target platform in crossbitness scenarios (i.e. x64_arm). Note that usually sizeof(X) != sizeof(Y).</s>

It's been proven experimentally that this subset of changes is needed in order to get outputs of x64_arm and arm_arm crossgens to be binary identical for S.P.C.dll and CoreFx assemblies.</s>

This PR also includes other related changes:
1. `LOG2_PTRSIZE` should depend on `_TARGET_64BIT_`;
2. more careful pointer arithmetic in `FakePromote` and `ReportPointersFromValueType`.

<s>This PR is to initiate design discussion (at alternatives discussion) and not a final work.</s>

**Update:** the final solution is to replace identified `sizeof` expressions with target-specifics constants

@jkotas PTAL

### CODE REVIEW

File: src/vm/ceeload.cpp [hunk 1]
```diff
@@ -2139,7 +2139,7 @@ void Module::BuildStaticsOffsets(AllocMemTracker *pamTracker)
                         (S_SIZE_T(2 * sizeof(DWORD))*(S_SIZE_T(dwNumTypes)+S_SIZE_T(1)))));
 
                 for (DWORD i = 0; i < dwIndex; i++) {
-                    pRegularStaticOffsets[i * 2    ] = dwGCHandles[0]*sizeof(OBJECTREF);
```
Comment (jkotas) [inline src/vm/ceeload.cpp]: `sizeof(OBJECTREF)` can be just `TARGET_POINTER_SIZE`. We use `TARGET_POINTER_SIZE` for object reference sizes in many places already, so a few more are not a concern.

File: src/vm/crossbitness.h [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+
+    struct ArrayBase : public Object
+    {
+        DWORD m_NumComponents;
+#ifdef _TARGET_64BIT_
+        DWORD pad;
+#endif // _TARGET_64BIT_
+    };
+
+    struct StringObject : public Object
```
Comment (jkotas) [inline src/vm/crossbitness.h]: I do not see this used.

File: src/vm/crossbitness.h [hunk 3]
```diff
@@ -0,0 +1,46 @@
+#ifndef CROSSBITNESS_H
+#define CROSSBITNESS_H
+
+struct CrossBitness
```
Comment (jkotas) [inline src/vm/crossbitness.h]: I would rather use explicitly manually defined sizes than trust the C++ compiler that it is going to layout the structures the way we expect.

Building upon approach I took in https://github.com/dotnet/coreclr/pull/17876/files .

File: src/vm/crossbitness.h (hunk 2 above)
Comment (jkotas) [reply src/vm/crossbitness.h]: I guess it was needed before https://github.com/dotnet/coreclr/pull/17876/

Comment (echesakovMSFT) [reply src/vm/crossbitness.h]: Yes, you're right. It was needed before for ObjSizeOf(StringObject) in StringObject::GetSize.

File: src/vm/ceeload.cpp (hunk 1 above)
Comment (echesakovMSFT) [reply src/vm/ceeload.cpp]: Sure, I will replace with `TARGET_POINTER_SIZE` instead

File: src/vm/crossbitness.h (hunk 3 above)
Comment (echesakovMSFT) [reply src/vm/crossbitness.h]: @jkotas I also though about this alternative. But then I would also need to replace all ObjSizeOf(Object) and ObjSizeOf(ArrayBase) expressions with some constants as well. 

Is any standard name for `sizeof(Object) + sizeof(ObjHeader)`?

**Comment by jkotas** [comment]
> `sizeof(Object) + sizeof(ObjHeader)`

I would call it `OBJECT_BASE_SIZE`.

**Comment by jkotas** [comment]
Or `Object::GetBaseSize()`

**Comment by echesakovMSFT** [comment]
@jkotas The x64_arm crossgen version in https://github.com/echesakovMSFT/coreclr/tree/CrossCrossGen compiles and runs (at least on Windows) and I can crossgen `S.P.C.dll` and all CoreFx assemblies

**Comment by echesakovMSFT** [comment]
@jkotas Regarding placements such constants - would asmconstants.h be appropriate file? Or I can keep them in crossbitness.h (for 32 and 64 bit targets) ?

**Comment by jkotas** [comment]
`asmconstants.h` is per architecture, so it would need to be duplicated. 

I would probably just add them at the top of object.h before `MIN_OBJECT_SIZE`, something like:

```
#define OBJHEADER_SIZE      TARGET_POINTER_SIZE /* ObjHeader */

#define OBJECT_SIZE         TARGET_POINTER_SIZE /* MethodTable* */
#define OBJECT_BASESIZE     (OBJHEADER_SIZE + OBJECT_SIZE)

#define ARRAYBASE_SIZE      (OBJECT_SIZE /* PTR_MethodTable */ + TARGET_POINTER_SIZE /* _numComponents + pad */)
#define ARRAYBASE_BASESIZE  (OBJHEADER_SIZE + ARRAYBASE_SIZE)

#define MIN_OBJECT_SIZE     (2*TARGET_POINTER_SIZE + OBJHEADER_SIZE)
```

File: src/vm/object.inl [hunk 4]
```diff
@@ -257,7 +257,7 @@ inline /* static */ unsigned ArrayBase::GetBoundsOffset(MethodTable* pMT)
     if (!pMT->IsMultiDimArray()) 
         return(offsetof(ArrayBase, m_NumComponents));
     _ASSERTE(pMT->GetInternalCorElementType() == ELEMENT_TYPE_ARRAY);
```
Comment (jkotas) [inline src/vm/object.inl]: You can also fix `offsetof(ArrayBase, m_NumComponents)` a few lines above while you are on it. Replace it with `OBJECT_SIZE /* offset(ArrayBase, m_NumComponents */` ?

Comment (echesakovMSFT) [reply src/vm/object.inl]: Done. Thanks for catching this

**Comment by echesakovMSFT** [comment]
After I replaced `ObjSizeOf` and `CrossBitness_SizeOf` with predefined constants I think next logical step should be adding some kind of checking (either at compile time or running time) that the values of these constants stay valid. For example,
```
C_ASSERT(sizeof(ObjHeader) == OBJHEADER_SIZE);
C_ASSERT(sizeof(Object) == OBJECT_SIZE);
```
However, such assertions would be valid only for non-crossbitness scenarios (so I can limit them to such scenarois only with ifdefs).

@jkotas  What do you think?

**Comment by jkotas** [comment]
You can add these static asserts somewhere if you would like.

I am not really concerned about these particular constants. This is unlikely to change; and if it does change and the constants get out of sync by accident, it will be easy to notice because of nothing will work without it.

**Comment by echesakovMSFT** [comment]
@jkotas Agree, this assertions would not be that helpful. I rebased and squashed some commits and wait for green ci

**Pull request merged**